### PR TITLE
InputStream refactoring in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR updates the InputStream method names and simplifies the generated code for tagged parameters.

### CODE REVIEW

File: csharp/src/Ice/InputStream.cs [hunk 1]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
-        {
-            float[] value = new float[ReadAndCheckSeqSize(4)];
-            ReadNumericArray(value);
-            return value;
-        }
+        /// <summary>Reads a sequence from the stream.</summary>
+        /// <param name="minElementSize">The minimum size of each element of the sequence, in bytes.</param>
+        /// <param name="reader">The input stream reader used to read each element of the sequence.</param>
+        /// <returns>The sequence read from the stream as an ICollection{T}. The return value does not fully implement
+        /// ICollection{T}, in particular you can only call GetEnumerator() once on this collection.</returns>
```
Comment (pepone) [inline csharp/src/Ice/InputStream.cs:310]: I find the comment misleading, it doesn't return the sequence read from the stream (only the size was read at the point the methods return), the collection is later read upon enumeration and that is why `GetEnumerator` should only be called once.

File: csharp/src/Ice/InputStream.cs [hunk 2]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
         {
-            int sz = ReadSize();
-            var dict = new SortedDictionary<TKey, TValue>();
-            for (int i = 0; i < sz; ++i)
+            if (ReadTaggedParamHeader(tag, EncodingDefinitions.TagFormat.FSize))
             {
-                TKey key = keyReader(this);
-                TValue value = valueReader(this);
-                dict.Add(key, value);
+                _ = ReadInt(); // skip byte size
```
Comment (pepone) [inline csharp/src/Ice/InputStream.cs:545]: the comment seems bogus, size is a int32 here

File: csharp/src/Ice/InputStream.cs [hunk 3]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
+        public T[]? ReadTaggedVariableSizeElementArray<T>(int tag, int minElementSize, InputStreamReader<T> reader)
         {
-            if (ReadOptional(tag, OptionalFormat.FSize))
+            if (ReadTaggedVariableSizeElementSequence(tag, minElementSize, reader) is ICollection<T> collection)
             {
-                Skip(4);
-                return ReadStringArray();
+                var array = new T[collection.Count];
+                collection.CopyTo(array, 0);
+                return array;
```
Comment (pepone) [inline csharp/src/Ice/InputStream.cs:571]: Replace with Linq ToArray

File: csharp/src/Ice/InputStream.cs [hunk 4]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
-            if (size < 0 || size > _buffer.Count - _pos)
+            int size = ReadSize(buffer, encoding);
+            if (size == 0)
             {
-                throw new IndexOutOfRangeException($"cannot skip {size} bytes");
+                return "";
+            }
+            else
+            {
+                return _utf8.GetString(buffer.AsSpan(size < 254 ? 1 : 4, size));
```
Comment (pepone) [inline csharp/src/Ice/InputStream.cs:704]: I think there is a bug here from my previous InputStream PR that introduced this static method

```suggestion
                return _utf8.GetString(buffer.AsSpan(size < 255 ? 1 : 5, size));
```

File: cpp/src/slice2cs/CsUtil.cpp [hunk 5]
```diff
@@ -1154,33 +1169,45 @@ Slice::CsGenerator::sequenceUnmarshalCode(const SequencePtr& seq, const string&
     {
         out << "(" << serializable << ") " << stream << ".ReadSerializable()";
     }
-    else if(generic.empty())
+    else if (generic.empty())
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply cpp/src/slice2cs/CsUtil.cpp:1172]: We should use the same parenthesis style everywhere - and this style is not C#-specific.

File: csharp/src/Ice/InputStream.cs [hunk 6]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
+                                                                     InputStreamReader<T> reader)
         {
-            if (ReadOptional(tag, OptionalFormat.VSize))
+            if (ReadTaggedParamHeader(tag, EncodingDefinitions.TagFormat.VSize))
             {
-                SkipSize();
-                return ReadDoubleArray();
+                if (elementSize > 1)
+                {
+                    _ = ReadSize(); // skip byte size
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:443]: I meant `skip size that represents the size of this tagged parameter in bytes`.

File: csharp/src/Ice/InputStream.cs [hunk 7]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
+        /// <param name="elementSize">The size of each element, in bytes.</param>
+        /// <param name="reader">The input stream reader used to read each element of the sequence.</param>
+        /// <returns>The sequence read from the stream as an array, or null.</returns>
+        public T[]? ReadTaggedFixedSizeElementArray<T>(int tag, int elementSize, InputStreamReader<T> reader)
+        {
+            if (ReadTaggedFixedSizeElementSequence(tag, elementSize, reader) is ICollection<T> collection)
+            {
+                var array = new T[collection.Count];
+                collection.CopyTo(array, 0);
+                return array;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:422]: Linq would take an IEnumerable, cast it to ICollection to get the count, and then create this array? I think this more straightforward code highlights that Count is available and used to create an array with the correct size.

File: csharp/src/Ice/InputStream.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:443]: Fixed.

File: csharp/src/Ice/InputStream.cs [hunk 8]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
-            if (ReadOptional(tag, OptionalFormat.VSize))
+            if (ReadTaggedParamHeader(tag, EncodingDefinitions.TagFormat.VSize))
             {
-                SkipSize();
-                return ReadDoubleArray();
+                if (elementSize > 1)
+                {
+                    _ = ReadSize(); // skip byte size
+                }
+                return ReadSequence<T>(elementSize, reader);
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:445]: Fixed.

File: csharp/src/Ice/InputStream.cs [hunk 9]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
             }
+            return dict;
         }
 
-        /// <summary>
-        /// Extracts a float value from the stream.
-        /// </summary>
-        /// <returns>The extracted float.</returns>
-        public float ReadFloat()
+        /// <summary>Reads an enum value from the stream; this method does not validdate the value.</summary>
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:272]: Fixed.

File: csharp/src/Ice/InputStream.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:704]: Fixed.

File: csharp/src/Ice/InputStream.cs [hunk 10]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
+                                                                        int minElementSize,
+                                                                        InputStreamReader<T> reader)
         {
-            var ident = new Identity(this);
-            if (ident.Name.Length == 0)
+            if (ReadTaggedParamHeader(tag, EncodingDefinitions.TagFormat.FSize))
             {
-                return null;
+                _ = ReadInt(); // skip byte size
+                return ReadSequence<T>(minElementSize, reader);
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:591]: Fixed.

File: csharp/src/Ice/InputStream.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/InputStream.cs:422]: fair enough 

The linq implementation is basically the same plus the cast

https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs#L22

File: csharp/src/Ice/InputStream.cs [hunk 11]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
             return value;
         }
 
-        public static long ReadLong(ReadOnlySpan<byte> buffer) => BitConverter.ToInt64(buffer);
+        /// <summary>Reads a short from the stream.</summary>
+        /// <returns>The short read from the stream.</returns>
+        public short ReadShort()
+        {
+            Debug.Assert(_buffer.Array != null);
+            short value = BitConverter.ToInt16(_buffer.Array, _buffer.Offset + _pos);
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:216]: Fixed, you meant `AsSpan`.

File: csharp/src/Ice/InputStream.cs [hunk 12]
```diff
@@ -592,456 +208,526 @@ public long ReadLong()
...
         {
             Debug.Assert(_buffer.Array != null);
-            float value = BitConverter.ToSingle(_buffer.Array, _buffer.Offset + _pos);
-            _pos += 4;
+            T[] value = new T[ReadAndCheckSeqSize(elementSize)];
+            int byteCount = Buffer.ByteLength(value);
+            Debug.Assert(_buffer.Count - _pos >= byteCount);
+            Buffer.BlockCopy(_buffer.Array, _buffer.Offset + _pos, value, 0, byteCount);
+            _pos += byteCount;
             return value;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs:287]: Fixed as suggested, also replaced parameter elementSize by `Unsafe.SizeOf<T>` as "hinted" by your comment.

File: csharp/src/Ice/InputStream.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/InputStream.cs]: Fixed.

**Pull request merged**